# SPDX-License-Identifier: Apache-2.0 OR ISC OR MIT
diff --git a/Makefile b/Makefile
index abc1234..def5678 100644
--- a/Makefile
+++ b/Makefile
@@ -100,7 +100,7 @@ hol.sh: pa_j.cmo ${HOLSRC} bignum.cmo hol_loader.cmo update_database.ml
 	if [ `uname` = "Linux" ] || [ `uname` = "Darwin" ] ; then \
 		if [ ${OCAML_UNARY_VERSION} = "5" ] || [ ${OCAML_VERSION} = "4.14" ] ; then \
-			ocamlfind ocamlmktop -package zarith -o ocaml-hol zarith.cma bignum.cmo hol_loader.cmo ; \
+			ocamlfind ocamlmktop -package zarith,findlib -o ocaml-hol zarith.cma bignum.cmo hol_loader.cmo ; \
 			sed "s^__DIR__^`pwd`^g; s^__USE_MODULE__^$(HOLLIGHT_USE_MODULE)^g" hol_4.14.sh > hol.sh ; \
 		else \
 			ocamlmktop -o ocaml-hol nums.cma bignum.cmo hol_loader.cmo ; \
